rflx/pyrflx/message.py

```diff
@@ -0,0 +1,168 @@
+
+    @property
+    def binary(self) -> bytes:
+        bits = "".join([f.typeval.binary for f in self._fields[1:]])
+        if len(bits) % 8:
+            raise ValueError(f"message length must be dividable by 8 ({len(bits)})")
+        return b"".join(
+            [int(bits[i : i + 8], 2).to_bytes(1, "big") for i in range(0, len(bits), 8)]
+        )
+
+    @property
+    def fields(self) -> List[str]:
+        return [f.name for f in self._model.all_fields[1:-1]]
+
+    @property
+    def accessible_fields(self) -> List[str]:
+        fields = [f.name for f in self._fields]
+        added = True
+        while added:
+            added = False
+            for l in self._model.outgoing(model.Field(fields[-1])):
+                if (
+                    l.target.name not in fields
+                    and l.target != model.FINAL
+                    and l.condition.simplified(self.__field_values) == TRUE
+                    and (
+                        l.length == UNDEFINED
+                        or isinstance(l.length.simplified(self.__field_values), Number)
+                    )
+                ):
+                    fields.append(l.target.name)
+                    added = True
+                    break
+        return fields[1:]
+
+    @property
+    def set_fields(self) -> List[str]:
+        return [f.name for f in self._fields[1:]]
+
+    @property
+    def final(self) -> bool:
+        for o in self._model.outgoing(model.Field(self._fields[-1].name)):
+            if o.target == model.FINAL and o.condition.simplified(self.__field_values) == TRUE:
+                return True
+        return False
+
+    @property
+    def __field_values(self) -> Mapping[Name, Expr]:
+        return {
+            **{
+                Variable(v.name): v.typeval.expr
+                for v in self._fields
+                if isinstance(v.typeval, ScalarValue) and v.typeval.initialized
+            },
+            **{Length(v.name): v.length for v in self._fields},
+            **{First(v.name): v.first for v in self._fields},
+        }
+
+    def __add_field(
+        self, name: str, typeval: TypeValue, first: Expr, length: Expr = UNDEFINED
+    ) -> None:
+        fld = Field(name)
+        fld.typeval = typeval
+        fld.first = first
+        if isinstance(fld.typeval, ScalarValue):
+            fld.length = Number(fld.typeval.size)
+        else:
+            fld.length = length.simplified(self.__field_values)
+            if not isinstance(fld.length, Number):
+                raise RuntimeError(f"unable to resolve length: {repr(fld.length)}")
+            if isinstance(fld.typeval, OpaqueValue) and fld.length.value != fld.typeval.length:
+                raise ValueError(f"invalid data length: {fld.length.value} != {fld.typeval.length}")
+        for f in self._fields:
+            if fld.name == f.name:
+                self._fields = self._fields[: self._fields.index(f)]
+                break
+        self._fields.append(fld)
```

Comment:
The current implementation invalidates all successor fields. It would be better to allow changing fields (e.g. addresses) without invalidating the rest of the message. A dict that contains all fields of the message and their current status and successor, could be more suitable. When a field value is changed, the conditions of all successor fields have to be checked again.

Reply:
Fixed by https://github.com/Componolit/RecordFlux/pull/158/commits/949c3c3aeccce59600e906a5d351d16f8de3d6ea.